File: balanced-braces.java
import java.util.*;
class Solution{
	
	public static void main(String []argh)
	{
		Scanner sc = new Scanner(System.in);
		
        String openingBrace = "[{(";
        String closingBrace = "]})";
        
		while (sc.hasNext()) {
			
			String input=sc.next();
            Stack<Character> stack = new Stack<Character>();
            boolean isBalanced = true;
            for(int i=0; i<input.length();i++)
            {
                Character c = input.charAt(i);
                
                if(openingBrace.contains(c.toString()))
                    stack.push(c);
                else
                    if(stack.empty())
                    {
                        isBalanced = false;
                        break;
                    }
                else
                    if(stack.peek().equals(openingBrace.charAt(closingBrace.lastIndexOf(c))))
                        stack.pop();
					else
					{
						isBalanced=false;
						break;
					}
            }
            
            if(stack.empty () && isBalanced)
                System.out.println("true");
            else
                System.out.println("false");
		}
	}
}

File: balanced-braces.cs
using System;
using System.Collections.Generic;

namespace ConsoleAppBalancedBrackets
{
    class Program
    {
        static void Main(string[] args)
        {
            int count = Convert.ToInt32(Console.ReadLine());

            for (int i = 0; i < count; i++)
            {
                string input = Console.ReadLine();
                Console.WriteLine(isBalanced(input));
            }
        }

        static string isBalanced(string input)
        {
            string bracesOpening = "({[";
            string bracesClosing = ")}]";

            string result = string.Empty;

            Stack<char> stackBraces = new Stack<char>();

            foreach (char c in input)
            {
                if (bracesOpening.Contains(c.ToString()))
                {
                    stackBraces.Push(c);
                }
                else
                if (stackBraces.Count == 0)
                {
                    result = "NO";
                    break;
                }
                else
                if (stackBraces.Peek() != bracesOpening[bracesClosing.IndexOf(c)])
                {
                    result = "NO";
                    break;
                }
                else
                    stackBraces.Pop();
            }

            return stackBraces.Count > 0 || result == "NO" ? "NO" : "YES";
        }
    }
}
